python/ray/tune/integration/kubernetes.py

```diff
@@ -0,0 +1,151 @@
+import kubernetes
+import subprocess
+
+from ray import services, logger
+from ray.autoscaler.command_runner import KubernetesCommandRunner
+from ray.tune.syncer import NodeSyncer
+from ray.tune.sync_client import SyncClient
+
+
+def NamespacedKubernetesSyncer(namespace):
+    """
+    Wrapper to return a ``KubernetesSyncer`` for a specific
+    Kubernetes namespace.
+
+    Args:
+        namespace (str): Kubernetes namespace.
+
+    Returns: A ``KubernetesSyncer`` class to be passed to
+        ``tune.run(sync_to_driver)``.
+
+    Example:
+
+        .. code-block:: python
+
+        from ray.tune.integration.kubernetes import NamespacedKubernetesSyncer
+        tune.run(train,
+                 sync_to_driver=NamespacedKubernetesSyncer("ray"))
+
+    """
+
+    class _NamespacedKubernetesSyncer(KubernetesSyncer):
+        _namespace = namespace
+
+    return _NamespacedKubernetesSyncer
+
+
+class KubernetesSyncer(NodeSyncer):
+    """
+    KubernetesSyncer used for synchronization of checkpoints between
+    Kubernetes pods.
+
+    This syncer extends the node syncer, but is usually instantiated
+    without a custom sync client. The sync client defaults to
+    ``KubernetesSyncClient`` instead.
+
+    KubernetesSyncer uses the default namespce ``ray``. You should
+    probably use ``NamespacedKubernetesSyncer`` to return a class
+    with a custom namespace instead.
+    """
+
+    _namespace = "ray"
+
+    def __init__(self, local_dir, remote_dir, sync_client=None):
+        self.local_ip = services.get_node_ip_address()
+        self.local_node = self._get_kubernetes_node_by_ip(self.local_ip)
+        self.worker_ip = None
+        self.worker_node = None
+
+        sync_client = sync_client or KubernetesSyncClient(
+            namespace=self.__class__._namespace)
+
+        super(NodeSyncer, self).__init__(local_dir, remote_dir, sync_client)
+
+    def set_worker_ip(self, worker_ip):
+        self.worker_ip = worker_ip
+        self.worker_node = self._get_kubernetes_node_by_ip(worker_ip)
+
+    def _get_kubernetes_node_by_ip(self, node_ip):
+        """Return node name by internal or external IP"""
+        kubernetes.config.load_incluster_config()
+        api = kubernetes.client.CoreV1Api()
+        pods = api.list_namespaced_pod(self._namespace)
+        for pod in pods.items:
+            if pod.status.host_ip == node_ip or \
+               pod.status.pod_ip == node_ip:
+                return pod.metadata.name
+
+        logger.error(
+            "Could not find Kubernetes node name for IP {}".format(node_ip))
```

Comment:
```suggestion
            "Could not find Kubernetes pod name for IP {}".format(node_ip))
```

Reply:
It's sort of weird that we're using the pod IP here, that's not something you'd usually do in k8s-land. Not a big deal and I assume this is just to match the ray node resource but might want to use pod name in the error messages where possible at least

Reply:
Yes, that's here for compatibility reasons as Tune syncs using the node IPs. I'll look into this if there's another way, but I guess that would require refactoring some distributed trial management code.